pkg/solana/client/multinode_client.go

```diff
@@ -0,0 +1,307 @@
+	chStopInFlight chan struct{}
+
+	chainInfoLock sync.RWMutex
+	// intercepted values seen by callers of the rpcClient excluding health check calls. Need to ensure MultiNode provides repeatable read guarantee
+	highestUserObservations mn.ChainInfo
+	// most recent chain info observed during current lifecycle (reseted on DisconnectAll)
+	latestChainInfo mn.ChainInfo
+}
+
+func NewMultiNodeClient(endpoint string, cfg *config.TOMLConfig, requestTimeout time.Duration, log logger.Logger) (*MultiNodeClient, error) {
+	client, err := NewClient(endpoint, cfg, requestTimeout, log)
+	if err != nil {
+		return nil, err
+	}
+
+	return &MultiNodeClient{
+		Client:         *client,
+		cfg:            cfg,
+		subs:           make(map[mn.Subscription]struct{}),
+		chStopInFlight: make(chan struct{}),
+	}, nil
+}
+
+// registerSub adds the sub to the rpcClient list
+func (m *MultiNodeClient) registerSub(sub mn.Subscription, stopInFLightCh chan struct{}) error {
+	m.subsSliceMu.Lock()
+	defer m.subsSliceMu.Unlock()
+	// ensure that the `sub` belongs to current life cycle of the `rpcClient` and it should not be killed due to
+	// previous `DisconnectAll` call.
+	select {
+	case <-stopInFLightCh:
+		sub.Unsubscribe()
+		return fmt.Errorf("failed to register subscription - all in-flight requests were canceled")
+	default:
+	}
+	// TODO: BCI-3358 - delete sub when caller unsubscribes.
+	m.subs[sub] = struct{}{}
+	return nil
+}
+
+func (m *MultiNodeClient) Dial(ctx context.Context) error {
+	// Not relevant for Solana as the RPCs don't need to be dialled.
+	return nil
+}
+
+func (m *MultiNodeClient) SubscribeToHeads(ctx context.Context) (<-chan *Head, mn.Subscription, error) {
+	ctx, cancel, chStopInFlight, _ := m.acquireQueryCtx(ctx, m.cfg.TxTimeout())
+	defer cancel()
+
+	pollInterval := m.cfg.MultiNode.PollInterval()
+	if pollInterval == 0 {
+		return nil, nil, errors.New("PollInterval is 0")
+	}
+	timeout := pollInterval
+	poller, channel := mn.NewPoller[*Head](pollInterval, m.LatestBlock, timeout, m.log)
+	if err := poller.Start(ctx); err != nil {
+		return nil, nil, err
+	}
+
+	err := m.registerSub(&poller, chStopInFlight)
+	if err != nil {
+		poller.Unsubscribe()
+		return nil, nil, err
+	}
+
+	return channel, &poller, nil
+}
+
+func (m *MultiNodeClient) SubscribeToFinalizedHeads(ctx context.Context) (<-chan *Head, mn.Subscription, error) {
+	ctx, cancel, chStopInFlight, _ := m.acquireQueryCtx(ctx, m.contextDuration)
+	defer cancel()
+
+	finalizedBlockPollInterval := m.cfg.MultiNode.FinalizedBlockPollInterval()
+	if finalizedBlockPollInterval == 0 {
+		return nil, nil, errors.New("FinalizedBlockPollInterval is 0")
+	}
+	timeout := finalizedBlockPollInterval
+	poller, channel := mn.NewPoller[*Head](finalizedBlockPollInterval, m.LatestFinalizedBlock, timeout, m.log)
+	if err := poller.Start(ctx); err != nil {
+		return nil, nil, err
+	}
+
+	err := m.registerSub(&poller, chStopInFlight)
+	if err != nil {
+		poller.Unsubscribe()
+		return nil, nil, err
+	}
```

Comment:
does this poller capture every new block on solana? or does it just poll intermittently and fetch the latest every X seconds?

we will likely run into performance issues if we are capturing _every_ new block

Reply:
It doesn't capture every block, just every X seconds which we can specify in the TOML configs.